src/plugins/markdown.ts

```diff
@@ -5,67 +5,156 @@
      */
-    public compile(markdownFiles: IFile[], { renderBlock }: ICompiler) {
-        const pageStore: Map<string, IPageData> = new Map();
+    public compile(markdownFiles: IFile[], compiler: ICompiler) {
+        const pageStore = this.buildPageMap(markdownFiles, compiler);
+        // nav must be generated before pages because it rewrites references
+        const nav = this.buildNavTree(pageStore);
+        const pages = pageStore.toObject();
+        return { nav, pages };
+    }
+
+    private buildPageMap(markdownFiles: IFile[], { renderBlock }: ICompiler) {
+        const pageStore: PageMap = new PageMap();
         markdownFiles
             .map((file) => {
-                const absolutePath = file.path;
-                const fileContents = file.read();
-                const { content, metadata, renderedContent } = renderBlock(fileContents);
-                const page = makePage({
-                    absolutePath,
+                const { content, metadata, renderedContent } = renderBlock(file.read());
+                return pageStore.add({
+                    absolutePath: file.path,
                     contentRaw: content,
                     contents: renderedContent,
                     metadata,
                 });
-                const ref = page.reference;
-                if (pageStore.has(ref)) {
-                    console.warn(`Found duplicate reference "${ref}"; overwriting previous data.`);
-                    console.warn("Rename headings or use metadata `reference` key to disambiguate.");
-                }
-                pageStore.set(ref, page);
-                return page;
             })
             .map((page) => {
-                if (page.contents) {
-                    const newContent = page.contents.reduce((array, content) => {
-                        if (typeof content === "string" || content.tag !== "include") {
-                            return array.concat(content);
-                        }
-                        const pageToInclude = pageStore.get(content.value as string);
-                        if (pageToInclude === undefined) {
-                            throw new Error(`Unknown @include reference '${content.value}' in '${page.reference}'`);
-                        }
-                        return array.concat(pageToInclude.contents!);
-                    }, [] as StringOrTag[]);
-                    page.contents = newContent;
-                }
+                // using `reduce` so we can add one or many entries for each node
+                page.contents = page.contents.reduce((array, content) => {
+                    if (typeof content === "string" || content.tag !== "include") {
+                        return array.concat(content);
+                    }
+                    // inline @include page
+                    try {
+                        const pageToInclude = pageStore.remove(content.value);
```

Comment:
These shouldn't be removed because of the reason i mentioned. Furthermore, they're not going to show up in the nav if there isn't a `@page` reference to them